File: Qt e Progetti/Progetti/2017-2018 Progetto Calcolatrice/Polygonator - Menon e Dalla Via/C++/Model/PolygonFactory.cpp
#include "PolygonFactory.h"

template<class T>
T *PolygonFactory::checkPolygon(const Polygon &polygon) {
    if (T::canBeCreatedWith(polygon)) {
        return new T(polygon.getVertices());
    }

    return nullptr;
}

Polygon *PolygonFactory::makePolygon(const list<Point> &vertices) {
    Polygon polygon = Polygon(vertices);
    list<std::function<Polygon *(const Polygon &)>> functionList;

    functionList.emplace_back(checkPolygon<Triangle>);
    functionList.emplace_back(checkPolygon<Square>);
    functionList.emplace_back(checkPolygon<Rhombus>);
    functionList.emplace_back(checkPolygon<Rectangle>);
    functionList.emplace_back(checkPolygon<Kite>);
    functionList.emplace_back(checkPolygon<Parallelogram>);
    functionList.emplace_back(checkPolygon<Quadrilateral>);
    functionList.emplace_back(checkPolygon<Pentagon>);
    functionList.emplace_back(checkPolygon<Hexagon>);

    for (const auto &function : functionList) {
        Polygon *p = function(polygon);

        if (p) {
            return p;
        }
    }

    return new Polygon(vertices);
}

File: Qt e Progetti/Progetti/2017-2018 Progetto Calcolatrice/Polygonator - Menon e Dalla Via/Java/src/Model/PolygonFactory.java
package Model;

import Model.Polygons.*;

import java.util.Arrays;
import java.util.List;

public abstract class PolygonFactory {

    private static <T extends Polygon> T checkPolygon(final Polygon polygon, Class<T> clazz) throws Exception {
        if ((boolean)clazz.getDeclaredMethod("canBeCreatedWith", Polygon.class).invoke(clazz, polygon)) {
            return clazz.getDeclaredConstructor(List.class).newInstance(polygon.getVertices());
        }

        return null;
    }

    public static Polygon makePolygon(List<Point> vertices) throws Exception {
        Polygon polygon = new Polygon(vertices);

        List<Class<? extends Polygon>> list = Arrays.asList(
                Triangle.class, Square.class, Rhombus.class,
                Rectangle.class, Kite.class, Parallelogram.class, Quadrilateral.class,
                Pentagon.class, Hexagon.class
        );

        for (Class<? extends Polygon> clazz : list) {
            Polygon obj = checkPolygon(polygon, clazz);

            if (obj != null) {
                return obj;
            }
        }
        return polygon;
    }
}
